lucet-runtime/lucet-runtime-internals/src/lock_testpoints.rs

```diff
@@ -0,0 +1,117 @@
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::time::Duration;
+
+pub struct SyncWaiter {
+    arrived: Arc<AtomicBool>,
+    proceed: Arc<AtomicBool>,
+}
+
+impl SyncWaiter {
+    pub fn wait(&self) {
+        while !self.arrived.load(Ordering::SeqCst) {
+            std::thread::sleep(Duration::from_millis(10));
+        }
+
+        self.proceed.store(true, Ordering::SeqCst);
+    }
+
+    pub fn wait_and_then<U, F: FnOnce() -> U>(&self, f: F) -> U {
+        while !self.arrived.load(Ordering::SeqCst) {
+            std::thread::sleep(Duration::from_millis(10));
+        }
+
+        let res = f();
+
+        self.proceed.store(true, Ordering::SeqCst);
+
+        res
+    }
+}
+
+pub struct Syncpoint {
```

Comment:
Do you think you could add some `///` comments briefly explaining what a sync point is, and how we use this in tests?

Reply:
We could link to the docs you wrote in #495 btw, I don't think this has to go into the details of `KillSwitch` so much as elaborating on how this helps force certain race conditions